network/gossip/libp2p/topology/topicBasedTopology.go

```diff
@@ -6,66 +6,172 @@ import (
+	}
 
-		// extract ids of subscribers to the topic
-		subscribers = ids.Filter(filter.HasRole(roles...))
+	if len(myFanout) == 0 {
+		return nil, fmt.Errorf("topology size reached zero")
 	}
+	return myFanout, nil
+}
+
+// subsetChannel returns a random subset of the identity list that is passed. `shouldHave` represents set of
+// identities that should be included in the returned subset.
+// Returned identities should all subscribed to the specified `channel`.
+// Note: this method should not include identity of its executor.
+func (t *TopicBasedTopology) subsetChannel(ids flow.IdentityList, shouldHave flow.IdentityList,
+	channel string) (flow.IdentityList, error) {
+	if _, ok := engine.IsClusterChannelID(channel); ok {
+		return t.clusterChannelHandler(ids, shouldHave)
+	} else {
+		return t.nonClusterChannelHandler(ids, shouldHave, channel)
+	}
+}
+
+// subsetRole returns a random subset of the identity list that is passed. `shouldHave` represents set of
+// identities that should be included in the returned subset.
+// Returned identities should all be of one of the specified `roles`.
+// Note: this method should not include identity of its executor.
+func (t TopicBasedTopology) subsetRole(ids flow.IdentityList, shouldHave flow.IdentityList, roles flow.RoleList) (flow.IdentityList, error) {
+	if shouldHave != nil {
+		// excludes irrelevant roles from should have set
+		shouldHave = shouldHave.Filter(filter.HasRole(roles...))
 
-	// excluding the node itself from its topology
-	subscribers = subscribers.Filter(filter.Not(filter.HasNodeID(t.me)))
+		// excludes the node itself from should have set
+		shouldHave = shouldHave.Filter(filter.Not(filter.HasNodeID(t.me)))
+	}
+
+	// excludes the node itself from all
+	ids = ids.Filter(filter.Not(filter.HasNodeID(t.me)))
 
-	// samples subscribers of a connected graph
-	subscriberSample, _ := connectedGraphSample(subscribers, t.seed)
+	sample, err := t.sampleConnectedGraph(ids, shouldHave)
+	if err != nil {
+		return nil, fmt.Errorf("could not sample a connected graph: %w", err)
+	}
 
-	return subscriberSample, nil
+	return sample, nil
+}
+
+// sampleConnectedGraph receives two lists: all and shouldHave. It then samples a connected fanout
+// for the caller that includes the shouldHave set. Independent invocations of this method over
+// different nodes, should create a connected graph.
+// Fanout is the set of nodes that this instance should get connected to in order to create a
+// connected graph.
+func (t TopicBasedTopology) sampleConnectedGraph(all flow.IdentityList, shouldHave flow.IdentityList) (flow.IdentityList, error) {
+
+	if len(all) == 0 {
+		return nil, fmt.Errorf("empty identity list")
+	}
+
+	if len(shouldHave) == 0 {
+		// choose (n+1)/2 random nodes so that each node in the graph will have a degree >= (n+1) / 2,
+		// guaranteeing a connected graph.
+		size := uint(LinearFanoutFunc(len(all)))
+		return all.DeterministicSample(size, t.seed), nil
+
+	} else {
```

Comment:
```suggestion
	}
```

Reply:
don't need the `else` since the `if` has a return.

Reply:
fixed in https://github.com/onflow/flow-go/pull/153/commits/bdf52fece1e8e8f6a5fa3c320648c15d0c316795